File: 3161-block-placement-queries/solution.cpp
#include <vector>
#include <algorithm>
#include <set>

using namespace std;

class SegmentTree {
private:
    int n;
    vector<int> tree;

public:
    SegmentTree(int n) {
        this->n = n;
        tree.assign(4 * n, 0);
    }

    void update(int node, int start, int end, int idx, int val) {
        if (start == end) {
            tree[node] = val;
            return;
        }
        int mid = start + (end - start) / 2;
        if (idx <= mid) {
            update(2 * node, start, mid, idx, val);
        } else {
            update(2 * node + 1, mid + 1, end, idx, val);
        }
        tree[node] = max(tree[2 * node], tree[2 * node + 1]);
    }

    int query(int node, int start, int end, int l, int r) {
        if (r < start || end < l) return 0;
        if (l <= start && end <= r) return tree[node];
        int mid = start + (end - start) / 2;
        return max(query(2 * node, start, mid, l, r),
                   query(2 * node + 1, mid + 1, end, l, r));
    }
};

class Solution {
public:
    vector<bool> getResults(vector<vector<int>>& queries) {
        int max_x = 0;
        for (const auto& q : queries) {
            max_x = max(max_x, q[1]);
        }

        SegmentTree st(max_x + 1);
        set<int> obstacles;
        obstacles.insert(0);
        vector<bool> results;

        for (const auto& q : queries) {
            if (q[0] == 1) {
                int x = q[1];
                auto it = obstacles.lower_bound(x);
                int prev_val = *prev(it);
                
                // Set the size of the new interval starting at prev_val ending at x
                st.update(1, 0, max_x, x, x - prev_val);
                
                // If there is an obstacle to the right of x, its interval shrinks
                if (it != obstacles.end()) {
                    int next_val = *it;
                    st.update(1, 0, max_x, next_val, next_val - x);
                }
                
                obstacles.insert(x);
            } else {
                int x = q[1];
                int sz = q[2];
                
                // Find the largest obstacle <= x
                auto it = obstacles.upper_bound(x);
                int prev_val = *prev(it);
                
                // Find the maximum gap inside [0, prev_val]
                int max_gap = st.query(1, 0, max_x, 0, prev_val);
                
                // The remaining gap is the space from prev_val to x
                int rem = x - prev_val;
                
                results.push_back(max(max_gap, rem) >= sz);
            }
        }

        return results;
    }
};

File: 3161-block-placement-queries/solution.py
import bisect

class SegmentTree:
    def __init__(self, n):
        self.n = n
        self.tree = [0] * (2 * n)
        
    def update(self, i, val):
        i += self.n
        self.tree[i] = val
        while i > 1:
            self.tree[i >> 1] = max(self.tree[i], self.tree[i ^ 1])
            i >>= 1
            
    def query(self, l, r):
        res = 0
        l += self.n
        r += self.n + 1
        while l < r:
            if l & 1:
                res = max(res, self.tree[l])
                l += 1
            if r & 1:
                r -= 1
                res = max(res, self.tree[r])
            l >>= 1
            r >>= 1
        return res

class Solution:
    def getResults(self, queries: list[list[int]]) -> list[bool]:
        max_x = 0
        for q in queries:
            max_x = max(max_x, q[1])
            
        st = SegmentTree(max_x + 1)
        obstacles = [0]
        results = []
        
        for q in queries:
            if q[0] == 1:
                x = q[1]
                idx = bisect.bisect_left(obstacles, x)
                prev_val = obstacles[idx - 1]
                
                # Set the size of the new interval starting at prev_val ending at x
                st.update(x, x - prev_val)
                
                # If there is an obstacle to the right of x, its interval shrinks
                if idx < len(obstacles):
                    next_val = obstacles[idx]
                    st.update(next_val, next_val - x)
                    
                obstacles.insert(idx, x)
            else:
                x, sz = q[1], q[2]
                idx = bisect.bisect_right(obstacles, x) - 1
                prev_val = obstacles[idx]
                
                # Find maximum gap within [0, prev_val]
                max_gap = st.query(0, prev_val)
                
                # Remaining space is the region from prev_val to x
                rem = x - prev_val
                
                results.append(max(max_gap, rem) >= sz)
                
        return results